### ISSUE TITLE
[Bug]: Window Embedding from WINCLIP is same across all patch

### ISSUE THREAD

**Issue by nadeemn** [openvinotoolkit/anomalib#2571]
### Describe the bug

I was implementing the **WINCLIP** paper with reference to this library, and came across with one issue. I was debugging using the MVTec Dataset to see how each window embedding is aligning with the text embedding. To my surprise each of the window embedding '**CLS**' tokens are same, leading to the same anomaly score across the window. This is not what the **WINCLIP** paper is portraying.
According to the paper, the window embedding '**CLS**' token aligns well with the text embedding, in that case, all the window should not throw the same pixel score or anomaly score.

PS: I was debugging in the Jupyter Notebook of the WinCLIP implementation and did not installed this library. 

### Dataset

MVTec

### Model

Other (please specify in the field below)

### Steps to reproduce the behavior

Steps to reproduce the error:
1. Extract the code for WINCLIP model.
2. Pass one image to the model.
3. Debug or print the _get_window_embedding function.
4. You can see the same 'CLS' embedding for each window. 
This should not be the case.

### OS information

OS information:
- OS: Linux
- Python version: 3.12.9
- PyTorch version: 2.6
- CUDA/cuDNN version: 12.6

### Expected behavior

Expected Behaviour:
Window embedding should be different for atleast some.

Current Behaviuor:
All window embedding are same leading to same score for all scales.

### Screenshots

I can only show the tensor output from my notebook.
The below tensor is for a window scale of 3.
```
tensor([[ 0.0788, -0.2007, -0.6241,  ..., -0.1215, -0.1220, -0.0777],
        [ 0.0788, -0.2007, -0.6241,  ..., -0.1215, -0.1220, -0.0777],
        [ 0.0788, -0.2007, -0.6241,  ..., -0.1215, -0.1220, -0.0777],
        ...,
        [ 0.0788, -0.2007, -0.6241,  ..., -0.1215, -0.1220, -0.0777],
        [ 0.0788, -0.2007, -0.6241,  ..., -0.1215, -0.1220, -0.0777],
        [ 0.0788, -0.2007, -0.6241,  ..., -0.1215, -0.1220, -0.0777]],
       device='cuda:0')
```

### Pip/GitHub

pip

### Configuration YAML

```yaml
NA
```

### Logs

```shell
NA
```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by nadeemn** [comment]
Unable to upload the screenshot. Copying the tensor output below:


tensor([[ 0.0788, -0.2007, -0.6241,  ..., -0.1215, -0.1220, -0.0777],
        [ 0.0788, -0.2007, -0.6241,  ..., -0.1215, -0.1220, -0.0777],
        [ 0.0788, -0.2007, -0.6241,  ..., -0.1215, -0.1220, -0.0777],
        ...,
        [ 0.0788, -0.2007, -0.6241,  ..., -0.1215, -0.1220, -0.0777],
        [ 0.0788, -0.2007, -0.6241,  ..., -0.1215, -0.1220, -0.0777],
        [ 0.0788, -0.2007, -0.6241,  ..., -0.1215, -0.1220, -0.0777]],
       device='cuda:0')

**Comment by djdameln** [comment]
Hello, I'm having some trouble reproducing your issue. Can you please share the line numbers and/or print statements that you used to obtain your debugging output?

**Comment by nadeemn** [comment]
Hello,
Sure, go to /src/models/image/winclip/torch_model.py file.
Create an instance for Winclip class. Pass an input image to the model (I used mvtec dataset, so a anomaly one).
Print out the tensors under _get_window_embeddings function. (Line 243).
Print out the masked tensor after this line of code:
masked = torch.cat([torch.index_select(feature_map, 1, mask) for mask in masks]) ------- (Line 262)
You can see that the masked tensor contains window patch of different scales, each assosciated with a class embedding token.

Now, look at the output after the transformer layer or after projecting the global pooling operation. i.e., after line 276
Each CLS token corresponding to the window is same, leading to same score for all the patch windows. (Contradicting the paper)

I am unable to upload a screenshot. But please find the tensor output below:

tensor([[ 0.0788, -0.2007, -0.6241, ..., -0.1215, -0.1220, -0.0777],
[ 0.0788, -0.2007, -0.6241, ..., -0.1215, -0.1220, -0.0777],
[ 0.0788, -0.2007, -0.6241, ..., -0.1215, -0.1220, -0.0777],
...,
[ 0.0788, -0.2007, -0.6241, ..., -0.1215, -0.1220, -0.0777],
[ 0.0788, -0.2007, -0.6241, ..., -0.1215, -0.1220, -0.0777],
[ 0.0788, -0.2007, -0.6241, ..., -0.1215, -0.1220, -0.0777]],
device='cuda:0')

**Comment by djdameln** [comment]
Hi, thanks for specifying. I had a closer look at the internals of the torch model, and as far as I can tell, the `_get_window_embeddings` method correctly extracts the window embeddings from the feature map.

> Print out the masked tensor after this line of code:
masked = torch.cat([torch.index_select(feature_map, 1, mask) for mask in masks]) ------- (Line 262)
You can see that the masked tensor contains window patch of different scales, each assosciated with a class embedding token.

The `masked` tensor has a shape of `[num_windows, patches_per_window + 1, dimensionality]`. The +1 is there because the class token is prepended to each of the window embeddings. Prepending the class token is a common practice in vision transformers as it helps aggregate the global, high-level features of the image. The value of the prepended class embedding is the same for all images, or in this case windows. Its value is learned during the contrastive pre-training stage of the CLIP model, and fixed during inference.

> Now, look at the output after the transformer layer or after projecting the global pooling operation. i.e., after line 276
Each CLS token corresponding to the window is same, leading to same score for all the patch windows. (Contradicting the paper)

I'm not sure exactly what you mean by this. The `pooled` tensor has a shape of `[num_windows, dimensionality]`. The patch dimension has been removed as a result of the global pooling operation, thus obtaining a single feature vector for each window location. This single feature vector contains the aggregated information of the class token and the patch tokens. In the next steps, we will obtain an anomaly score for each window location by computing the similarity to the text embeddings. This is in line with the paper, and inspecting the values of the `pooled` tensor shows a unique feature vector for each of the window locations.

**Comment by djdameln** [comment]
Thanks for clarifying. I understand your issue now, but I'm not able to repoduce it. When I inspect the `pooled` tensor in my local environment, I'm seeing a unique feature vector for every window location. Here's the code that I used:

```python
from anomalib.models import WinClip
from torchvision.io import read_image
from torchvision.transforms.v2.functional import to_dtype
import torch

# create model
model = WinClip()

# read image
image_path = "datasets/MVTec/hazelnut/test/crack/000.png"
image = read_image(image_path)
image = to_dtype(image, torch.float32, scale=True).unsqueeze(0)

# apply input transforms
image = model.pre_processor.transform(image)

# predict
model(image)

```
And this is the output of the `pooled` tensor for scale 3:
```
tensor([[-0.1722, -0.0507, -0.5750,  ...,  0.0702, -0.1176, -0.2373],
        [-0.1756, -0.0469, -0.5754,  ...,  0.0712, -0.1213, -0.2388],
        [-0.1768, -0.0433, -0.5733,  ...,  0.0712, -0.1237, -0.2367],
        ...,
        [-0.1964, -0.0504, -0.5586,  ...,  0.0617, -0.1098, -0.2280],
        [-0.1846, -0.0484, -0.5650,  ...,  0.0677, -0.1152, -0.2318],
        [-0.1822, -0.0533, -0.5678,  ...,  0.0697, -0.1140, -0.2293]])

```

I'm using the latest Anomalib version from the main branch of the repository.

If you share the Anomalib version and the code that was used, I can try to reproduce the issue on my side.

**Comment by nadeemn** [comment]
Hi, 
Thankyou for the clarification. I didnt cloned or installed the library. For my thesis, I was playing a little bit around WinCLIP code, and found out my mistake.
I didnt normalized my input tensor to the specific value according to the underline CLIP model. It worked now.

Thankyou again for your time. You can close the issue. :)
